cpp/dolfinx/io/utils.cpp

```diff
@@ -0,0 +1,526 @@
+        std::move(recv_buffer), std::move(recv_values_buffer), shape);
+  };
+  // NOTE: src and dest are transposed here because we're reversing the
+  // direction of communication
+  const auto [entities_data_b, entities_values, shape_eb]
+      = candidate_ranks(comm, compound_type, nodes_g_p, recv_disp, dest, src,
+                        entitiesp, std::span(entitiesp_v));
+  mdspan_t<const std::int64_t, 2> entities_data(entities_data_b.data(),
+                                                shape_eb);
+
+  // -- E. From the received (key, value) data, determine which keys
+  //    (entities) are on this process.
+  //
+  // TODO: We have already received possibly tagged entities from other
+  //       ranks, so we could use the received data to avoid creating
+  //       the std::map for *all* entities and just for candidate
+  //       entities.
+  auto select_entities
+      = [](const mesh::Topology& topology, auto xdofmap,
+           std::span<const std::int64_t> nodes_g,
+           std::span<const int> cell_vertex_dofs, auto entities_data,
+           std::span<const T> entities_values)
+  {
+    spdlog::info("XDMF build map");
+    auto c_to_v = topology.connectivity(topology.dim(), 0);
+    if (!c_to_v)
+      throw std::runtime_error("Missing cell-vertex connectivity.");
+
+    std::map<std::int64_t, std::int32_t> input_idx_to_vertex;
+    for (int c = 0; c < c_to_v->num_nodes(); ++c)
+    {
+      auto vertices = c_to_v->links(c);
+      std::span xdofs(xdofmap.data_handle() + c * xdofmap.extent(1),
+                      xdofmap.extent(1));
+      for (std::size_t v = 0; v < vertices.size(); ++v)
+        input_idx_to_vertex[nodes_g[xdofs[cell_vertex_dofs[v]]]] = vertices[v];
+    }
+
+    std::vector<std::int32_t> entities;
+    std::vector<T> data;
+    std::vector<std::int32_t> entity(entities_data.extent(1));
+    for (std::size_t e = 0; e < entities_data.extent(0); ++e)
+    {
+      bool entity_found = true;
+      for (std::size_t i = 0; i < entities_data.extent(1); ++i)
+      {
+        if (auto it = input_idx_to_vertex.find(entities_data(e, i));
+            it == input_idx_to_vertex.end())
+        {
+          // As soon as this received index is not in locally owned
+          // input global indices skip the entire entity
+          entity_found = false;
+          break;
+        }
+        else
+          entity[i] = it->second;
+      }
+
+      if (entity_found)
+      {
+        entities.insert(entities.end(), entity.begin(), entity.end());
+        data.push_back(entities_values[e]);
+      }
+    }
+
+    return std::pair(std::move(entities), std::move(data));
+  };
+
+  MPI_Type_free(&compound_type);
+
+  return select_entities(topology, xdofmap, nodes_g, cell_vertex_dofs,
+                         entities_data, std::span(entities_values));
+}
+//-----------------------------------------------------------------------------
+/// @cond
+template std::pair<std::vector<std::int32_t>, std::vector<double>>
```

Comment:
Explicit instantiation can be handy for a (very) small set of types. This is a bit over-the-top now, so I would suggest moving the implementation into the header file.